sdr_airspy.c

```diff
@@ -0,0 +1,456 @@
+    uint32_t slen;
+    unsigned next_free_buffer;
+    unsigned free_bufs;
+    unsigned block_duration;
+    static int dropping = 0;
+    static uint64_t sampleCounter = 0;
+
+    if (Modes.exit) {
+        airspy_stop_rx(AIRSPY.dev);
+        return 0;
+    }
+
+    // Lock the data buffer variables before accessing them
+    pthread_mutex_lock(&Modes.data_mutex);
+
+    next_free_buffer = (Modes.first_free_buffer + 1) % MODES_MAG_BUFFERS;
+    outbuf = &Modes.mag_buffers[Modes.first_free_buffer];
+    lastbuf = &Modes.mag_buffers[(Modes.first_free_buffer + MODES_MAG_BUFFERS - 1) % MODES_MAG_BUFFERS];
+    free_bufs = (Modes.first_filled_buffer - next_free_buffer + MODES_MAG_BUFFERS) % MODES_MAG_BUFFERS;
+
+    pthread_mutex_unlock(&Modes.data_mutex);
+
+    // Give the demodulater what it expects
+    soxr_process(AIRSPY.resampler, inptr, i_len, &i_done, outptr, MODES_RTL_BUF_SIZE, &o_done);
+    for (i = 0; i < o_done; i++)
+        AIRSPY.airspy_bytes[i] = (int8_t)(outptr[i] >> 4) + 127;
+
+    slen = o_done;
+
+    if (free_bufs == 0 || (dropping && free_bufs < MODES_MAG_BUFFERS/2)) {
+        // FIFO is full. Drop this block.
+        dropping = 1;
+        outbuf->dropped += slen;
+        sampleCounter += slen;
+        return 0;
+    }
+
+    dropping = 0;
+
+    // Compute the sample timestamp and system timestamp for the start of the block
+    outbuf->sampleTimestamp = sampleCounter + 12e6 / Modes.sample_rate;
+    sampleCounter += slen;
+    block_duration = 1e9 * slen / Modes.sample_rate;
+
+    // Get the approx system time for the start of this block
+    clock_gettime(CLOCK_REALTIME, &outbuf->sysTimestamp);
+    outbuf->sysTimestamp.tv_nsec -= block_duration;
+    normalize_timespec(&outbuf->sysTimestamp);
+
+    // Copy trailing data from last block (or reset if not valid)
+    if (outbuf->dropped == 0) {
+        memcpy(outbuf->data, lastbuf->data + lastbuf->length, Modes.trailing_samples * sizeof(uint16_t));
+    } else {
+        memset(outbuf->data, 0, Modes.trailing_samples * sizeof(uint16_t));
+    }
+
+    // Convert the new data
+    outbuf->length = slen;
+    AIRSPY.converter(AIRSPY.airspy_bytes, &outbuf->data[Modes.trailing_samples], slen, AIRSPY.converter_state, &outbuf->mean_level, &outbuf->mean_power); 
+
+    // Push the new data to the demodulation thread
+    pthread_mutex_lock(&Modes.data_mutex);
+
+    Modes.mag_buffers[next_free_buffer].dropped = 0;
+    Modes.mag_buffers[next_free_buffer].length = 0;  // just in case
+    Modes.first_free_buffer = next_free_buffer;
+
+    // Accumulate CPU while holding the mutex, and restart measurement
+    end_cpu_timing(&airspy_thread_cpu, &Modes.reader_cpu_accumulator);
+    start_cpu_timing(&airspy_thread_cpu);
+
+    pthread_cond_signal(&Modes.data_cond);
+    pthread_mutex_unlock(&Modes.data_mutex);
+
+    return (0);
+}
+
+void airspyRun()
+{
+    if (!AIRSPY.dev) {
+        return;
+    }
+
+    start_cpu_timing(&airspy_thread_cpu);
```

Comment:
This is unlikely to work as expected because the airspy callbacks are happening in a different thread, not the thread calling airspyRun()